snips_nlu/intent_parser/lookup_intent_parser.py

```diff
@@ -0,0 +1,403 @@
+        self.map = dict()
+        entity_placeholders = _get_entity_placeholders(dataset, self.language)
+
+        for (key, val) in self.generate_io_mapping(
+                dataset[INTENTS], entity_placeholders
+        ):
+            # handle key collisions -*- remove ambiguous entries -*-
+            if key in self.map and self.map[key] != val:
+                self.map.pop(key)
+            else:
+                self.map[key] = val
+
+        return self
+
+    @log_result(
+        logger, logging.DEBUG,
+        "DictDeterministicIntentParser result -> {result}"
+    )
+    @log_elapsed_time(logger, logging.DEBUG, "Parsed in {elapsed_time}.")
+    @fitted_required
+    def parse(self, text, intents=None, top_n=None):
+        """Performs intent parsing on the provided *text*
+
+        Intent and slots are extracted simultaneously through pattern matching
+
+        Args:
+            text (str): input
+            intents (str or list of str): if provided, reduces the scope of
+                intent parsing to the provided list of intents
+            top_n (int, optional): when provided, this method will return a
+                list of at most top_n most likely intents, instead of a single
+                parsing result.
+                Note that the returned list can contain less than ``top_n``
+                elements, for instance when the parameter ``intents`` is not
+                None, or when ``top_n`` is greater than the total number of
+                intents.
+
+        Returns:
+            dict or list: the most likely intent(s) along with the extracted
+            slots. See :func:`.parsing_result` and :func:`.extraction_result`
+            for the output format.
+
+        Raises:
+            NotTrained: when the intent parser is not fitted
+        """
+        if isinstance(intents, str):
+            intents = [intents]
+
+        builtin_entities = self.builtin_entity_parser.parse(text,
+                                                            use_cache=True)
+        custom_entities = self.custom_entity_parser.parse(text, use_cache=True)
+        all_entities = builtin_entities + custom_entities
+
+        all_entities = deduplicate_overlapping_entities(all_entities)
+
+        placeholder_fn = lambda entity_name: _get_entity_name_placeholder(
+            entity_name, self.language
+        )
+        _, processed_text = replace_entities_with_placeholders(
+            text, all_entities, placeholder_fn
+        )
+
+        cleaned_processed_text = self._preprocess_text(processed_text)
+        cleaned_text = self._preprocess_text(text).lower()
+        cleaned_text = "".join(
+            [t for t in tokenize_light(cleaned_text, self.language)])
+
+        key = "".join(
+            [
+                t
+                for t in tokenize_light(cleaned_processed_text, self.language)
+                if normalize(t) not in self.stop_words
+            ]
+        ).lower()
```

Comment:
`cleaned_processed_text` and `cleaned_text` should not be post-processed,`self._preprocess_text` should return a valid key candidate for the map (cf comment below on the `_preprocess_text` method)